icu4c/source/common/uloc.cpp

```diff
@@ -1381,6 +1385,14 @@ _getVariant(const char* localeID,
         }
         if (pEnd != nullptr) { *pEnd = localeID; }
     }
+    int32_t variantsLength = localeID - start;
```

Comment:
If there is a very large amount of variants, that now are about to be rejected, the code will have spent a lot of effort repeatedly calling `sink->Append()` to copy all those characters before reaching this point. As the purpose of this PR is to avoid long running time, you might want to consider adding a preprocessing step at the beginning of this function, a simpler loop that just counts the length and returns early with an error if the variants are too long, before starting to copy anything (see `_getLanguage()` for an analogous such loop in one of the other of these functions).